indexing-service/src/test/java/io/druid/indexing/common/task/AppenderatorDriverRealtimeIndexTaskTest.java

```diff
@@ -0,0 +1,1312 @@
+    statusFuture.get();
+  }
+
+  @Test(timeout = 60_000L)
+  public void testBasics() throws Exception
+  {
+    expectPublishedSegments(1);
+    final AppenderatorDriverRealtimeIndexTask task = makeRealtimeTask(null);
+    final TaskToolbox taskToolbox = makeToolbox(task, tempFolder.newFolder());
+    final ListenableFuture<TaskStatus> statusFuture = runTask(task, taskToolbox);
+
+    // Wait for firehose to show up, it starts off null.
+    while (task.getFirehose() == null) {
+      Thread.sleep(50);
+    }
+
+    final TestFirehose firehose = (TestFirehose) task.getFirehose();
+
+    firehose.addRows(
+        ImmutableList.of(
+            ImmutableMap.of("t", now.getMillis(), "dim1", "foo", "met1", "1"),
+            ImmutableMap.of("t", now.getMillis(), "dim2", "bar", "met1", 2.0)
+        )
+    );
+
+    // Stop the firehose, this will drain out existing events.
+    firehose.close();
+
+    // Wait for publish.
+    Collection<DataSegment> publishedSegments = awaitSegments();
+
+    // Check metrics.
+    Assert.assertEquals(2, task.getMetrics().processed());
+    Assert.assertEquals(0, task.getMetrics().thrownAway());
+    Assert.assertEquals(0, task.getMetrics().unparseable());
+
+    // Do some queries.
+    Assert.assertEquals(2, sumMetric(task, null, "rows"));
+    Assert.assertEquals(3, sumMetric(task, null, "met1"));
+
+    awaitHandoffs();
+
+    for (DataSegment publishedSegment : publishedSegments) {
+      Optional<Map.Entry<SegmentDescriptor, Pair<Executor, Runnable>>> optional = handOffCallbacks.entrySet().stream()
+                                                                                                  .filter(e -> e.getKey().equals(new SegmentDescriptor(
+                                                                                                      publishedSegment.getInterval(),
+                                                                                                      publishedSegment.getVersion(),
+                                                                                                      publishedSegment.getShardSpec().getPartitionNum()
+                                                                                                  )))
+                                                                                                  .findFirst();
+
+      Assert.assertTrue(
+          publishedSegment + " missing from handoff callbacks: " + handOffCallbacks,
+          optional.isPresent()
+      );
+      Pair<Executor, Runnable> executorRunnablePair = optional.get().getValue();
+
+      // Simulate handoff.
+      executorRunnablePair.lhs.execute(executorRunnablePair.rhs);
+    }
+    handOffCallbacks.clear();
+
+    // Wait for the task to finish.
+    final TaskStatus taskStatus = statusFuture.get();
+    Assert.assertEquals(TaskState.SUCCESS, taskStatus.getStatusCode());
+  }
+
+  @Test(timeout = 60_000L)
+  public void testLateData() throws Exception
```

Comment:
Would you explain how this tests late arrivals?

Reply:
Sure.  It basically adds two input rows.  One "now" and one "now - 2 days".  It then tests that both events were added and the data is queryable

```
    firehose.addRows(
        ImmutableList.of(
            ImmutableMap.of("t", now.getMillis(), "dim1", "foo", "met1", "1"),
            // Data is from 2 days ago, should still be processed
            ImmutableMap.of("t", now.minus(new Period("P2D")).getMillis(), "dim2", "bar", "met1", 2.0)
        )
    );
// ....
    // Do some queries.
    Assert.assertEquals(2, sumMetric(task, null, "rows"));
    Assert.assertEquals(3, sumMetric(task, null, "met1"));

```